src/core/cuda/cuda_profile.cc

```diff
@@ -0,0 +1,465 @@
+        group, CUPTI_EVENT_GROUP_ATTR_EVENTS, &eventIdsSize, eventIds));
+
+    std::vector<uint64_t> values(numInstances);
+
+    for (uint32_t j = 0; j < numEvents; j++) {
+      values.clear();
+      valuesSize = sizeof(uint64_t) * numInstances;
+      TC_CUPTI_CHECK(cuptiEventGroupReadEvent(
+          group,
+          CUPTI_EVENT_READ_FLAG_NONE,
+          eventIds[j],
+          &valuesSize,
+          values.data()));
+      CHECK_EQ(numInstances, valuesSize / sizeof(uint64_t));
+      CHECK_LE(metricData.eventValues.size(), metricData.numEvents)
+          << "Too many events collected, metric expects only "
+          << metricData.numEvents;
+
+      // sum collect event values from all instances
+      auto sum = std::accumulate(
+          values.begin(), values.begin() + numInstances, uint64_t(0));
+
+      // normalize the event value to represent the total number of
+      // domain instances on the device
+      auto normalized = (sum * numTotalInstances) / numInstances;
+
+      metricData.eventIds.push_back(eventIds[j]);
+      metricData.eventValues.push_back(normalized);
+
+      if (FLAGS_cuda_profile_verbose_events) {
+        logEvent(
+            eventIds[j],
+            numInstances,
+            numTotalInstances,
+            sum,
+            normalized,
+            values.data());
+      }
+    }
+  }
+  for (uint64_t i = 0; i < metricData.eventGroups->numEventGroups; i++)
+    TC_CUPTI_CHECK(
+        cuptiEventGroupDisable(metricData.eventGroups->eventGroups[i]));
+}
+
+void CUPTIAPI getMetricValueCallback(
+    void* userdata,
+    CUpti_CallbackDomain domain,
+    CUpti_CallbackId cbid,
+    const CUpti_CallbackData* cbInfo) {
+  MetricData& metricData = *static_cast<MetricData*>(userdata);
+
+  // This callback is enabled only for launch so we shouldn't see
+  // anything else.
+  CHECK_EQ(cbid, CUPTI_DRIVER_TRACE_CBID_cuLaunchKernel)
+      << "Unexpected cbid: " << cbid;
+
+  // on entry, enable all the event groups being collected this pass,
+  // for metrics we collect for all instances of the event
+  if (cbInfo->callbackSite == CUPTI_API_ENTER) {
+    eventCollectionStart(metricData, cbInfo);
+  }
+
+  // on exit, read and record event values
+  if (cbInfo->callbackSite == CUPTI_API_EXIT) {
+    eventCollectionEnd(metricData);
+  }
+}
+CUpti_MetricValueKind getValueKind(const CudaMetric& metric) {
+  CUpti_MetricValueKind valueKind;
+  size_t valueKindSize = sizeof(valueKind);
+  TC_CUPTI_CHECK(cuptiMetricGetAttribute(
+      metric.id, CUPTI_METRIC_ATTR_VALUE_KIND, &valueKindSize, &valueKind));
+  return valueKind;
+}
+
+double metricValueAsDouble(const CudaMetric& metric) {
+  auto valueKind = getValueKind(metric);
+  if (valueKind == CUPTI_METRIC_VALUE_KIND_DOUBLE) {
+    return metric.value.metricValueDouble;
+  } else if (valueKind == CUPTI_METRIC_VALUE_KIND_PERCENT) {
+    return metric.value.metricValuePercent;
+  } else {
+    CHECK(false) << "Invalid metric value conversion.";
+  }
+}
```

Comment:
These functions result in warnings about control reaching
end of non-void function.

Reply:
The compiler can't figure out that `CHECK(false)` always leads to program termination. I could return dummy values to make the compiler happy.

Reply:
Or use something like `__builtin_unreachable`

Reply:
Please do.